File: src/sema.go
package src

import "fmt"

type SemaError struct {
	Line, Col int
	Message   string
}

func (e *SemaError) Error() string {
	return fmt.Sprintf("semantic error at %d:%d: %s", e.Line, e.Col, e.Message)
}

// Analyze validates the document AST and returns a SemaError on the first
// violation found.
func Analyze(doc *DocumentNode) error {
	if doc.Meta != nil {
		if err := checkMetaFields(doc.Meta); err != nil {
			return err
		}
	}
	for _, block := range doc.Blocks {
		if err := analyzeBlock(block); err != nil {
			return err
		}
	}
	return nil
}

func checkMetaFields(m *MetaNode) error {
	for _, required := range []string{"title"} {
		if _, ok := m.Fields[required]; !ok {
			return &SemaError{Line: m.Line, Col: m.Col, Message: fmt.Sprintf("meta block is missing required field %q", required)}
		}
	}
	return nil
}

func analyzeBlock(n Node) error {
	switch b := n.(type) {
	case *HeadingNode:
		if b.Level < 1 || b.Level > 6 {
			return &SemaError{Line: b.Line, Col: b.Col, Message: fmt.Sprintf("heading level %d out of range, must be 1-6", b.Level)}
		}
		return analyzeInlines(b.Children)
	case *ParagraphNode:
		return analyzeInlines(b.Children)
	case *CodeBlockNode:
		if b.RawCode == "" {
			return &SemaError{Line: b.Line, Col: b.Col, Message: "codeblock has empty content"}
		}
		return nil
	case *HRNode:
		return nil
	case *ListNode:
		if len(b.Items) == 0 {
			return &SemaError{Line: b.Line, Col: b.Col, Message: "list must contain at least one item"}
		}
		for _, item := range b.Items {
			if err := analyzeInlines(item.Children); err != nil {
				return err
			}
		}
		return nil
	case *QuoteNode:
		return analyzeInlines(b.Children)
	case *ImageNode:
		if b.Src == "" {
			return &SemaError{Line: b.Line, Col: b.Col, Message: "image is missing a source URL"}
		}
		return nil
	case *MetaNode:
		return &SemaError{Line: b.Line, Col: b.Col, Message: "meta block must appear only once, at document root"}
	default:
		return fmt.Errorf("semantic error: unknown node kind %q", n.Kind())
	}
}

func analyzeInlines(children []Node) error {
	for _, c := range children {
		inline, ok := c.(*InlineNode)
		if !ok {
			continue
		}
		switch inline.NodeKind {
		case InlineText, InlineCode:
			// leaf inline kinds, nothing further to validate
		case InlineBold, InlineItalic:
			if err := analyzeInlines(inline.Children); err != nil {
				return err
			}
		case InlineLink:
			if inline.URL == "" {
				return &SemaError{Line: inline.Line, Col: inline.Col, Message: "link is missing a URL"}
			}
			if err := analyzeInlines(inline.Children); err != nil {
				return err
			}
		default:
			return &SemaError{Line: inline.Line, Col: inline.Col, Message: fmt.Sprintf("unknown inline kind %q", inline.NodeKind)}
		}
	}
	return nil
}

File: src/sema.rs
use crate::ast::{
    BlockNode, DocumentNode, InlineKind, InlineNode, ItemChild, ListNode, MetaNode, QuoteChild,
    TableNode,
};
use crate::errors::SemaError;

/// Analyze validates the document AST and returns a SemaError on the first
/// semantic violation found[span_1](start_span)[span_1](end_span).
pub fn analyze(doc: &DocumentNode) -> Result<(), SemaError> {
    if let Some(ref meta) = doc.meta {
        check_meta_fields(meta)?;
    }
    for block in &doc.blocks {
        analyze_block(block)?;
    }
    Ok(())
}

fn check_meta_fields(m: &MetaNode) -> Result<(), SemaError> {
    for required in ["title"] {
        if !m.fields.contains_key(required) {
            return Err(SemaError::new(
                m.line,
                m.col,
                format!("meta block is missing required field \"{required}\""),
            ));
        }
    }
    Ok(())
}

fn analyze_block(block: &BlockNode) -> Result<(), SemaError> {
    match block {
        BlockNode::Heading(h) => {
            if h.level < 1 || h.level > 6 {
                return Err(SemaError::new(
                    h.line,
                    h.col,
                    format!("heading level {} out of range, must be 1-6", h.level),
                ));
            }
            analyze_inlines(&h.children)
        }
        BlockNode::Paragraph(p) => analyze_inlines(&p.children),
        BlockNode::CodeBlock(cb) => {
            if cb.raw_code.is_empty() {
                return Err(SemaError::new(
                    cb.line,
                    cb.col,
                    "codeblock has empty content",
                ));
            }
            Ok(())
        }
        BlockNode::HR(_) => Ok(()),
        BlockNode::List(l) => analyze_list(l),
        BlockNode::Quote(q) => analyze_quote_children(&q.children),
        BlockNode::Image(img) => {
            if img.src.is_empty() {
                return Err(SemaError::new(
                    img.line,
                    img.col,
                    "image is missing a source URL",
                ));
            }
            Ok(())
        }
        BlockNode::Table(t) => analyze_table(t),
        BlockNode::Callout(c) => analyze_inlines(&c.children),
        BlockNode::Raw(r) => {
            if r.html.is_empty() {
                return Err(SemaError::new(r.line, r.col, "raw block has empty content"));
            }
            Ok(())
        }
        BlockNode::Math(m) => {
            if m.latex.is_empty() {
                return Err(SemaError::new(
                    m.line,
                    m.col,
                    "math block has empty content",
                ));
            }
            Ok(())
        }
    }
}

fn analyze_list(list: &ListNode) -> Result<(), SemaError> {
    if list.items.is_empty() {
        return Err(SemaError::new(
            list.line,
            list.col,
            "list must contain at least one item",
        ));
    }
    for item in &list.items {
        analyze_item_children(&item.children)?;
    }
    Ok(())
}

fn analyze_table(t: &TableNode) -> Result<(), SemaError> {
    if t.rows.is_empty() {
        return Err(SemaError::new(
            t.line,
            t.col,
            "table must contain at least one row",
        ));
    }
    let mut width: Option<usize> = None;
    for row in &t.rows {
        if row.cells.is_empty() {
            return Err(SemaError::new(
                row.line,
                row.col,
                "row must contain at least one cell",
            ));
        }
        match width {
            None => width = Some(row.cells.len()),
            Some(w) if w != row.cells.len() => {
                return Err(SemaError::new(
                    row.line,
                    row.col,
                    format!(
                        "row has {} cells, expected {} to match the table's other rows",
                        row.cells.len(),
                        w
                    ),
                ));
            }
            _ => {}
        }
        for cell in &row.cells {
            analyze_inlines(&cell.children)?;
        }
    }
    Ok(())
}

fn analyze_item_children(children: &[ItemChild]) -> Result<(), SemaError> {
    for child in children {
        match child {
            ItemChild::List(nested) => analyze_list(nested)?,
            ItemChild::Inline(inline) => analyze_inline(inline)?,
        }
    }
    Ok(())
}

fn analyze_quote_children(children: &[QuoteChild]) -> Result<(), SemaError> {
    for child in children {
        match child {
            QuoteChild::Quote(nested) => analyze_quote_children(&nested.children)?,
            QuoteChild::Inline(inline) => analyze_inline(inline)?,
        }
    }
    Ok(())
}

fn analyze_inlines(children: &[InlineNode]) -> Result<(), SemaError> {
    for inline in children {
        analyze_inline(inline)?;
    }
    Ok(())
}

fn analyze_inline(inline: &InlineNode) -> Result<(), SemaError> {
    match &inline.kind {
        InlineKind::Text(_) | InlineKind::Code(_) | InlineKind::Math(_) => Ok(()),
        InlineKind::Bold(children)
        | InlineKind::Italic(children)
        | InlineKind::Strike(children) => analyze_inlines(children),
        InlineKind::Link { url, children } => {
            if url.is_empty() {
                return Err(SemaError::new(
                    inline.line,
                    inline.col,
                    "link is missing a URL",
                ));
            }
            analyze_inlines(children)
        }
    }
}
